README.md
[![Pub Version](https://badgen.net/pub/v/theme_generator_x)](https://pub.dev/packages/theme_generator_x/)
[![Dart SDK Version](https://badgen.net/pub/sdk-version/theme_generator_x)](https://pub.dev/packages/theme_generator_x/)
[![Pub popularity](https://badgen.net/pub/popularity/theme_generator_x)](https://pub.dev/packages/theme_generator_x/score)


# Theme generator

simple dart code generator that can help you to generate [flutter extension](https://api.flutter.dev/flutter/material/ThemeExtension-class.html) from JSON code with different formats

## Usage
### 1. install
you should install it to pub global as command line tool
```dart pub global activate theme_generator_x```

### 2. arguments
* `input` - path to your input JSON file
* `output` - path to your output `.dart` file
* `class_name` - name of generated extension class, e.g. **AppThemeDataColorsX**
* `use_dark` - flag if you need to generate dark colors too. **false** by default
* `keys_rename` - how to rename JSON keys
  
  `Available options`:
  * `camel_case` - **camelCase**
  * `original` - **original name**
  * `snake_case` - **snake_case**


## JSON structure
you can use different JSON key-value structures
### Allowed key-values
```json
{
    "primary": "#f6f4da",
    "secondary": "0xff9e9e9e",
    "some_color": "#ff000011"
}
```

array of colors
* `0 item` - light color
* `1 item` - dark color
```json
{
    "composite": [
        "#f6f4da",
        "0xff9e9e9e"
    ]
}
```

```json
{
    "test_color_simple": {
        "light": "#000011"
    },
    "test_color": {
        "dark": "#656213",
        "light": "#000011"
    }
}
```

also you can generate array of colors
```json
{
 "simple_map_of_array": {
        "light": [
            "#656213",
            "#000011"
        ]
    },
    "simple_map_of_array_dark": {
        "light": [
            "#995577",
            "#000011"
        ],
        "dark": [
            "#110022",
            "#000000"
        ]
    }
}
```

## CMD

```bash
theme_generator_x colors --input example/test_schema.json --output example/output.dart --class_name AppThemeDataColorsX
```

of course you can use absolute or relative paths to the files
```bash
theme_generator_x colors --input /Users/jWinterDay/Documents/app_theme_asset.json --output app_theme_color.dart --class_name AppThemeDataColorsX
```

## Flutter
now you have the extensions and can use it in the flutter app

```dart
MaterialApp(
    ...
    theme: ThemeData(
        useMaterial3: true,
    ).copyWith(
        extensions: <ThemeExtension<dynamic>>[
            lightAppThemeDataColorsXData(), // generated extension function
        ],
    ),
  ...
)
```

CHANGELOG.md
## 0.0.6
* light and dark color as list
## 0.0.5
* example
## 0.0.4
* example
## 0.0.3
* Updated README
## 0.0.2
* Make the package executable

## 0.0.1

* TODO: Describe initial release.


example/example.dart
// GENERATED CODE - DO NOT MODIFY BY HAND

import 'package:flutter/material.dart';

AppThemeDataColorsX lightAppThemeDataColorsXData() {
  return const AppThemeDataColorsX(
    primary: Color(0xfff6f4da),
    secondary: Color(0xff9e9e9e),
    someColor: Color(0xff000011),
    testColorSimple: Color(0xff000011),
    testColor: Color(0xff000011),
    simpleMapOfArray: <Color>[
      Color(0xff656213),
      Color(0xff000011),
    ],
    simpleMapOfArrayDark: <Color>[
      Color(0xff995577),
      Color(0xff000011),
    ],
    composite: Color(0xfff6f4da),
  );
}

AppThemeDataColorsX darkAppThemeDataColorsXData() {
  return const AppThemeDataColorsX(
    primary: Color(0xfff6f4da),
    secondary: Color(0xff9e9e9e),
    someColor: Color(0xff000011),
    testColorSimple: Color(0xff000011),
    testColor: Color(0xff656213),
    simpleMapOfArray: <Color>[
      Color(0xff656213),
      Color(0xff000011),
    ],
    simpleMapOfArrayDark: <Color>[
      Color(0xff110022),
      Color(0xff000000),
    ],
    composite: Color(0xff9e9e9e),
  );
}

@immutable
class AppThemeDataColorsX extends ThemeExtension<AppThemeDataColorsX> {
  const AppThemeDataColorsX({
    required this.primary,
    required this.secondary,
    required this.someColor,
    required this.testColorSimple,
    required this.testColor,
    required this.simpleMapOfArray,
    required this.simpleMapOfArrayDark,
    required this.composite,
  });

  final Color? primary;
  final Color? secondary;
  final Color? someColor;
  final Color? testColorSimple;
  final Color? testColor;
  final List<Color>? simpleMapOfArray;
  final List<Color>? simpleMapOfArrayDark;
  final Color? composite;

  @override
  AppThemeDataColorsX copyWith({
    Color? primary,
    Color? secondary,
    Color? someColor,
    Color? testColorSimple,
    Color? testColor,
    List<Color>? simpleMapOfArray,
    List<Color>? simpleMapOfArrayDark,
    Color? composite,
  }) {
    return AppThemeDataColorsX(
      primary: primary ?? this.primary,
      secondary: secondary ?? this.secondary,
      someColor: someColor ?? this.someColor,
      testColorSimple: testColorSimple ?? this.testColorSimple,
      testColor: testColor ?? this.testColor,
      simpleMapOfArray: simpleMapOfArray ?? this.simpleMapOfArray,
      simpleMapOfArrayDark: simpleMapOfArrayDark ?? this.simpleMapOfArrayDark,
      composite: composite ?? this.composite,
    );
  }

  @override
  AppThemeDataColorsX lerp(AppThemeDataColorsX? other, double t) {
    if (other is! AppThemeDataColorsX) {
      return this;
    }

    return AppThemeDataColorsX(
      primary: Color.lerp(primary, other.primary, t),
      secondary: Color.lerp(secondary, other.secondary, t),
      someColor: Color.lerp(someColor, other.someColor, t),
      testColorSimple: Color.lerp(testColorSimple, other.testColorSimple, t),
      testColor: Color.lerp(testColor, other.testColor, t),
      simpleMapOfArray: simpleMapOfArray,
      simpleMapOfArrayDark: simpleMapOfArrayDark,
      composite: Color.lerp(composite, other.composite, t),
    );
  }
}


lib/src/utils/utils.dart
// ignore_for_file: avoid_print, avoid_classes_with_only_static_members

// import 'package:figma/figma.dart' hide Document;

import 'package:recase/recase.dart';

class Utils {
  static const String _printColorBlack = '\x1B[30m';
  static const String _printColorRed = '\x1B[31m';
  static const String _printColorGreen = '\x1B[32m';
  static const String _printColorYellow = '\x1B[33m';
  static const String _printColorBlue = '\x1B[34m';
  static const String _printColorMagenta = '\x1B[35m';
  static const String _printColorCyan = '\x1B[36m';
  static const String _printColorWhite = '\x1B[37m';
  static const String _printColorReset = '\x1B[0m';

  static void printBlack(String str) => print('$_printColorBlack$str$_printColorReset');
  static void printRed(String str) => print('$_printColorRed$str$_printColorReset');
  static void printGreen(String str) => print('$_printColorGreen$str$_printColorReset');
  static void printYellow(String str) => print('$_printColorYellow$str$_printColorReset');
  static void printBlue(String str) => print('$_printColorBlue$str$_printColorReset');
  static void printMagenta(String str) => print('$_printColorMagenta$str$_printColorReset');
  static void printCyan(String str) => print('$_printColorCyan$str$_printColorReset');
  static void printWhite(String str) => print('$_printColorWhite$str$_printColorReset');
  static void printReset(String str) => print('$_printColorReset$str$_printColorReset');

  static String rename(String text, {required String to}) {
    return switch (to) {
      'camel_case' => ReCase(text).camelCase,
      'original' => text,
      'snake_case' => ReCase(text).snakeCase,
      _ => ReCase(text).camelCase,
    };
  }
}


lib/src/commands/colors_command.dart
import 'dart:io';

import 'package:args/command_runner.dart';
// import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;
import 'package:theme_generator_x/theme_generator_x.dart';

class ColorsCommand extends Command<void> {
  ColorsCommand() {
    argParser
      ..addOption(
        'input',
        abbr: 'p',
        help: 'colors JSON path',
      )
      ..addOption(
        'output',
        abbr: 'o',
        help: 'output file path',
      )
      ..addOption(
        'class_name',
        abbr: 'c',
        help: 'output class name',
      )
      ..addFlag(
        'use_dark',
        abbr: 'd',
      )
      ..addOption(
        'keys_rename',
        abbr: 'r',
        allowed: <String>[
          'camel_case', // camelCase
          'original',
          'snake_case', // snake_case
        ],
        defaultsTo: 'camel_case',
      );
  }

  @override
  String get description => 'theme generator x. colors';

  @override
  String get name => 'colors';

  @override
  String get invocation => '${runner?.executableName} $name [arguments] <directories>';

  // final Map<String, String> _fieldNameMap = <String, String>{};

  @override
  Future<void> run() async {
    Utils.printGreen('$name generator');

    // input path
    if (argResults?['input'] == null) {
      throw UsageException('Input file path must be not null', 'Enter color file path');
    }
    final String inputPath = argResults?['input'] as String;
    final String absInputPath = p.join(Directory.current.path, inputPath);
    Utils.printCyan('file input path: $absInputPath');
    final File inputFile = File(absInputPath);
    final bool inputExists = inputFile.existsSync();
    if (!inputExists) {
      throw UsageException('Input file does not exist', 'Ensure input file exists');
    }

    // output path
    if (argResults?['output'] == null) {
      throw UsageException('Output file path does not exist', 'Use correct output file path');
    }
    final String outputPath = argResults?['output'] as String;

    final String absOutputPath = p.join(Directory.current.path, outputPath);
    Utils.printCyan('file output path: $absOutputPath');
    final File outputFile = File(absOutputPath);
    final bool outputExists = outputFile.existsSync();
    if (!outputExists) {
      outputFile.createSync();
    }

    // use dark
    final bool useDark = argResults?['use_dark'] as bool;

    // class name
    if (argResults?['class_name'] == null) {
      throw UsageException('Color extension class name must be not null', 'Enter extension class name');
    }
    final String className = argResults?['class_name'] as String;

    final String keysRename = argResults?['keys_rename'] as String;

    // download files
    // final http.Response colorContent = await http.get(Uri.parse(colorPath));
    // if (colorContent.statusCode != 200) {
    //   throw UsageException(
    //     'Can not download color content ${colorContent.statusCode}',
    //     'Can not download color content',
    //   );
    // }

    // print(' colorContent.body = ${colorContent.body}');

    // final http.Response colorDarkContent = await http.get(Uri.parse(colorDarkUrl));
    // if (colorDarkContent.statusCode != 200) {
    //   throw UsageException(
    //     'Can not download color content ${colorDarkContent.statusCode}',
    //     'Can not download dark color content',
    //   );
    // }

    // generate
    // final StringBuffer sb = StringBuffer();
    // sb.writeln('// keysRename = $keysRename');
    // sb.writeln('// className = $className');
    // sb.writeln('// outputPath = $outputPath');
    // sb.writeln('// absOutputPath = $absOutputPath');

    //
    final ColorsUtils colorUtils = ColorsUtils();

    final String result = colorUtils.generateX(
      content: inputFile.readAsStringSync(),
      className: className,
      useDark: useDark,
      keysRename: keysRename,
    );

    outputFile.writeAsStringSync(result);
  }
}


lib/src/utils/colors_utils.dart
import 'dart:convert';
import 'package:theme_generator_x/theme_generator_x.dart';

const JsonDecoder _kDecoder = JsonDecoder();

const String _kExtensionTemplate = '''
@immutable
class #ClassName# extends ThemeExtension<#ClassName#> {
  const #ClassName# ({
    #Params#
  });

  #Fields#

  @override
  #ClassName# copyWith({
    #CopyWithArguments#
  }) {
    return #ClassName#(
      #CopyWithReturn#
    );
  }

  @override
  #ClassName# lerp(#ClassName#? other, double t) {
    if (other is! #ClassName#) {
      return this;
    }

    return #ClassName#(
      #LerpReturn#
    );
  }
}
''';

const String _kExtensionDataTemplate = '''
#ClassName# #DataMethodName# () {
  return const #ClassName#(
    #Values
  );
}
''';

class ColorsUtils {
  static const String _darkName = 'dark';
  static const String _lightName = 'light';

  /// #495aa6 => 0xff495aa6
  String replaceColorVal(String? color) {
    if (color == null) {
      throw Exception('color is null');
    }

    // 0xff9e9e9e
    // #495aa6 len = 7
    // #ff495aa6 len = 9
    if (!<int>[7, 9, 10].contains(color.length)) {
      throw Exception('Color [$color] does not have required format: #495aa6 or #ff495aa6');
    }

    // 0xff9e9e9e
    if (color.length == 10 && color.startsWith('0x')) {
      return 'Color($color)';
    }

    if (!color.startsWith('#')) {
      throw Exception('Color [$color] does not start with #');
      // return '0xffffffff';
    }

    // #495aa6
    if (color.length == 7) {
      final String val = color.replaceAll(RegExp('#'), '0xff');
      return 'Color($val)';
    }

    // #ff495aa6
    final String val = color.replaceAll(RegExp('#'), '0x');
    return 'Color($val)';
  }

  /// ```json
  /// {
  ///     "primary": "#f6f4da",
  ///     "secondary": "#656213",
  ///     "some_color": "#000011",
  ///     "test_color": {
  ///         "dark": "#656213",
  ///         "light": "#000011"
  ///     },
  ///     "storyLinearGradientBackground": {
  ///         "dark": [
  ///             "#656213",
  ///             "#000011"
  ///         ],
  ///         "light": [
  ///             "#656213",
  ///             "#000011"
  ///         ]
  ///     }
  /// }
  /// ```
  String generateX({
    required String content,
    required String className,
    required bool useDark,
    required String keysRename,
  }) {
    final StringBuffer sb = StringBuffer();

    sb.writeln('''
      // GENERATED CODE - DO NOT MODIFY BY HAND

      import 'package:flutter/material.dart';
    ''');

    final Map<String, dynamic> tokenMap = _kDecoder.convert(content) as Map<String, dynamic>;

    // result class
    final StringBuffer extParams = StringBuffer();
    final StringBuffer extFields = StringBuffer();
    final StringBuffer extCopyWithArguments = StringBuffer();
    final StringBuffer extCopyWithReturn = StringBuffer();
    final StringBuffer extLerpReturn = StringBuffer();

    // result data
    final StringBuffer dataLightValues = StringBuffer();
    final StringBuffer dataDarkValues = StringBuffer();

    // _kExtensionDataTemplate

    for (final MapEntry<String, dynamic> entry in tokenMap.entries) {
      final String keyResultName = Utils.rename(entry.key, to: keysRename);
      final dynamic value = entry.value;

      //
      extParams.writeln('required this.$keyResultName,');
      extCopyWithReturn.writeln('$keyResultName: $keyResultName ?? this.$keyResultName,');

      switch (value) {
        /// `1. simple string "#f6f4da"`
        case final String lVal:
          final String lightColorResult = replaceColorVal(lVal);
          // class
          extFields.writeln('final Color? $keyResultName;');
          extCopyWithArguments.writeln('Color? $keyResultName,');
          extLerpReturn.writeln('$keyResultName: Color.lerp($keyResultName, other.$keyResultName, t),');
          // data
          dataLightValues.writeln('$keyResultName: $lightColorResult,');
          if (useDark) {
            dataDarkValues.writeln('$keyResultName: $lightColorResult,');
          }

        /// `2. map {"light": "#f6f4da"} or {"light": "#f6f4da", "dark": "#000011"}`
        case {_lightName: final String lVal}:
          final String? dVal = value[_darkName];
          final String lightColorResult = replaceColorVal(lVal);
          final String darkColorResult = dVal == null ? lightColorResult : replaceColorVal(dVal);

          // class
          extFields.writeln('final Color? $keyResultName;');
          extCopyWithArguments.writeln('Color? $keyResultName,');
          extLerpReturn.writeln('$keyResultName: Color.lerp($keyResultName, other.$keyResultName, t),');
          // data
          dataLightValues.writeln('$keyResultName: $lightColorResult,');
          if (useDark) {
            dataDarkValues.writeln('$keyResultName: $darkColorResult,'); //
          }

        /// `3. map of array colors {"light": ["#656213", "#000011"], "dark": ["#656213", "#000011"]}. Dark is optional`
        case {_lightName: final dynamic lVal}:
          // light
          List<String> lightList = <String>[];
          if (lVal is List<dynamic>) {
            lightList = lVal.map((dynamic element) => replaceColorVal(element?.toString())).toList();
          }
          final String resultLigthStr = '''
            <Color>[
              ${lightList.join(',')},
            ]
          ''';

          // dark
          List<String>? darkList;
          String? resultDarkStr;
          if (useDark) {
            final dynamic dValRaw = value[_darkName];
            if (dValRaw is List<dynamic>) {
              darkList = dValRaw.map((dynamic element) => replaceColorVal(element?.toString())).toList();
            }
            resultDarkStr = darkList == null
                ? resultLigthStr
                : '''
                <Color>[
                  ${darkList.join(',')},
                ]
              ''';
          }

          // class
          extFields.writeln('final List<Color>? $keyResultName;');
          extCopyWithArguments.writeln('List<Color>? $keyResultName,');
          extLerpReturn.writeln('$keyResultName: $keyResultName,');
          // data
          dataLightValues.writeln('$keyResultName: $resultLigthStr,');
          if (useDark) {
            dataDarkValues.writeln('$keyResultName: $resultDarkStr,'); //
          }

        /// `4. list of strings ["#f6f4da", "0xff9e9e9e"]`
        case final List<dynamic> lVal: // final String lVal:

          final String lightColorResult = replaceColorVal(lVal[0].toString());

          // class
          extFields.writeln('final Color? $keyResultName;');
          extCopyWithArguments.writeln('Color? $keyResultName,');
          extLerpReturn.writeln('$keyResultName: Color.lerp($keyResultName, other.$keyResultName, t),');

          // data
          dataLightValues.writeln('$keyResultName: $lightColorResult,');
          if (useDark) {
            if (lVal.length != 2) {
              throw Exception('Unknown list of colors format $keyResultName: $entry');
            }

            final String darkColorResult = replaceColorVal(lVal[1]);

            dataDarkValues.writeln('$keyResultName: $darkColorResult,');
          }

        default:
          throw Exception('Unknown color format $keyResultName: $entry');
      }
    }

    // replace in template
    final String resultClassTemplate = _kExtensionTemplate
        .replaceAll(RegExp('#ClassName#'), className) // AppThemeDataColorsX
        .replaceAll(RegExp('#Params#'), extParams.toString())
        .replaceAll(RegExp('#Fields#'), extFields.toString())
        .replaceAll(RegExp('#CopyWithArguments#'), extCopyWithArguments.toString())
        .replaceAll(RegExp('#CopyWithReturn#'), extCopyWithReturn.toString())
        .replaceAll(RegExp('#LerpReturn#'), extLerpReturn.toString());

    final String resultLightDataTemplate = _kExtensionDataTemplate
        .replaceAll(RegExp('#ClassName#'), className) // AppThemeDataColorsX
        .replaceAll(RegExp('#DataMethodName#'), '$_lightName${className}Data') // ligthAppThemeDataColorsXData
        .replaceAll(RegExp('#Values'), dataLightValues.toString());

    String? resultDarkDataTemplate;
    if (useDark) {
      resultDarkDataTemplate = _kExtensionDataTemplate
          .replaceAll(RegExp('#ClassName#'), className) // AppThemeDataColorsX
          .replaceAll(RegExp('#DataMethodName#'), '$_darkName${className}Data') // darkAppThemeDataColorsXData
          .replaceAll(RegExp('#Values'), dataDarkValues.toString());
    }

    sb
      ..writeln(resultLightDataTemplate)
      ..writeln('\n')
      ..writeln(resultDarkDataTemplate ?? '')
      ..writeln('\n\n')
      ..writeln(resultClassTemplate);

    return sb.toString();
  }
}
